instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java

```diff
@@ -73,221 +57,71 @@ public void visitLabel(Label label) {
 
     super.visitLabel(label);
 
+    visitPossibleJump(label);
 
-    final boolean isJump = myEnumerator.isJump(label);
-    if (myLastJump != -1) {
-      Label l = new Label();
-
-      mv.visitVarInsn(Opcodes.ILOAD, getLineVariableNumber());
-      pushLineNumber(myLastLineJump);
-      mv.visitJumpInsn(Opcodes.IF_ICMPNE, l);
-
-      mv.visitVarInsn(Opcodes.ILOAD, getJumpVariableNumber());
-      mv.visitIntInsn(Opcodes.SIPUSH, myLastJump);
-      mv.visitJumpInsn(Opcodes.IF_ICMPNE, l);
-
-      touchLastJump();
-
-      if (isJump) {
-        Label l1 = new Label();
-        mv.visitJumpInsn(Opcodes.GOTO, l1);
-        mv.visitLabel(l);
-        mv.visitVarInsn(Opcodes.ILOAD, getJumpVariableNumber());
-        mv.visitJumpInsn(Opcodes.IFLT, l1);
-        touchBranch(true);
-        mv.visitLabel(l1);
-      }
-      else {
-        mv.visitLabel(l);
-      }
-    }
-    else if (isJump) {
-      mv.visitVarInsn(Opcodes.ILOAD, getJumpVariableNumber());
-      Label newLabelX = new Label();
-      mv.visitJumpInsn(Opcodes.IFLT, newLabelX);
-      touchBranch(true);
-      mv.visitLabel(newLabelX);
-    }
-
-
-    final Integer key = myEnumerator.getSwitchKey(label);
-    if (key != null) {
+    final LineEnumerator.Switch aSwitch = myEnumerator.getSwitch(label);
+    if (aSwitch != null) {
       mv.visitVarInsn(Opcodes.ALOAD, getCurrentClassDataNumber());
-      mv.visitVarInsn(Opcodes.ILOAD, getLineVariableNumber());
-      mv.visitVarInsn(Opcodes.ILOAD, getSwitchVariableNumber());
-      mv.visitIntInsn(Opcodes.SIPUSH, key);
+      pushIntValue(aSwitch.getLine());
+      pushIntValue(aSwitch.getIndex());
+      mv.visitIntInsn(Opcodes.SIPUSH, aSwitch.getKey());
       mv.visitMethodInsn(Opcodes.INVOKESTATIC, ProjectData.PROJECT_DATA_OWNER, "touchSwitch", "(Ljava/lang/Object;III)V", false);
     }
   }
 
-  private void touchBranch(final boolean trueHit) {
+  private void touchBranch(final boolean trueHit, final int jumpIndex, int line) {
     mv.visitVarInsn(Opcodes.ALOAD, getCurrentClassDataNumber());
-    mv.visitVarInsn(Opcodes.ILOAD, getLineVariableNumber());
-    mv.visitVarInsn(Opcodes.ILOAD, getJumpVariableNumber());
+    pushIntValue(line);
+    pushIntValue(jumpIndex);
     mv.visitInsn(trueHit ? Opcodes.ICONST_0 : Opcodes.ICONST_1);
     mv.visitMethodInsn(Opcodes.INVOKESTATIC, ProjectData.PROJECT_DATA_OWNER, "touchJump", "(Ljava/lang/Object;IIZ)V", false);
-
-    mv.visitIntInsn(Opcodes.SIPUSH, -1);
-    mv.visitVarInsn(Opcodes.ISTORE, getJumpVariableNumber());
-  }
-
-  private void touchLastJump() {
```

Comment:
if I remember correctly, this was used to hit false branch for cases like
```
if (true) {
  return xxx;
}
doSomth();
return zzz;
```

is it covered now?

Reply:
It is covered like this
```
if (true) {   // PARTIAL
  return xxx; // FULL
}
doSomth();    // NONE
return zzz;   // NONE
```
or with `false`
```
if (false) {   // PARTIAL
  return xxx; // NONE
}
doSomth();    // FULL
return zzz;   // FULL
```
Or if code is run twice with both `true/false`
```
if (b) {   // FULL
  return xxx; // FULL
}
doSomth();    // FULL
return zzz;   // FULL
```

Reply:
Oh no, I've asked how your code deal with this user sample, not about resulted coverage. Pan was not intended, sorry

Reply:
It should be covered as jumps (`IF_EQ label`) are transformed like this now:
```
IF_EQ true_label
GOTO false_label
true_label:
  touchTrue()
  GOTO label
false_label:
  touchFalse()
  // false code here
```

Reply:
The point of this example is the way of touching false branch when `doSomth` is called(but only if condition is false). Jump to it's label is changed to jump at internal label where touch is done, after that a jump to the original label is done. Therefore false touch is done only if this jump is done.

Reply:
Is it this code:
// a super call to filters may remove this jump, so myLastJump may be null after this line
         super.visitJumpInsn(opcode, trueLabel);
         super.visitJumpInsn(Opcodes.GOTO, falseLabel);
         super.visitLabel(trueLabel);  // true hit will be inserted here
         super.visitJumpInsn(Opcodes.GOTO, label);
         super.visitLabel(falseLabel); // false hit will be inserted here

Reply:
Yes, it is